<a href="https://colab.research.google.com/github/Ssonamy/MADSU/blob/main/Lab1.1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Импорт библиотек**



In [ ]:
import pandas as pd
import numpy as np
import kagglehub

# Download latest version

# Скачка датасета с пингвинами
path = kagglehub.dataset_download("satyajeetrai/palmer-penguins-dataset-for-eda")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'palmer-penguins-dataset-for-eda' dataset.
Path to dataset files: /kaggle/input/palmer-penguins-dataset-for-eda


## **Библиотека NumPy**

###**1. Сгенерируйте случайный целочисленный одномерный NymPy массив длинны 10, значения для которого берутся из диапазона от 0 до 50. Выполните следующие задания:**

  *   Выведите первые пять значений массива, выведите значения с третьего по медьмое включительно.

  *   Выведите уникальные значения и их количество.

  *   Отсортируйте значения по возрастанию, выведите индексы отсортированного массива.

  *   Найдите среднее, медианное, минимальное и максимальное значения.





Зафиксируем воспроизводимость результатов:

In [ ]:
np.random.seed(42)

Сгенерируйте случайный numpy-массив:

In [ ]:
a = np.random.randint(0, 50, 10)
a


array([38, 28, 14, 42,  7, 20, 38, 18, 22, 10])

Срез из первых пяти значений:

In [ ]:
a[:5]

array([38, 28, 14, 42,  7])

Срез с третьего по седьмомое значение:

In [ ]:
a[2:7]

array([14, 42,  7, 20, 38])

Уникальные значения в массиве:

In [ ]:
np.unique(a)

array([ 7, 10, 14, 18, 20, 22, 28, 38, 42])

Уникальные значения с числом вхождений:

In [ ]:
np.unique(a, return_counts=True)

(array([ 7, 10, 14, 18, 20, 22, 28, 38, 42]),
 array([1, 1, 1, 1, 1, 1, 1, 2, 1]))

Количество уникальных значений:

In [ ]:
len(np.unique(a))

9

Сортировка значений:

In [ ]:
np.sort(a)

array([ 7, 10, 14, 18, 20, 22, 28, 38, 38, 42])

Индексы отсортированного массива:

In [ ]:
np.argsort(a)

array([4, 9, 2, 7, 5, 8, 1, 0, 6, 3])

Среднее, медианное, минимальное и максимальное значения соответсвенно:

In [ ]:
a.mean(), np.median(a), a.min(), a.max()

(np.float64(23.7), np.float64(21.0), np.int64(7), np.int64(42))

###**2. Создайте массивы длины 16 со значениями между 30 и 60 с равномерным шагом.**


*   Превратите его в двумерный numpy-массив (вектор столбец)
*   Превратите его в двумерный numpy-массив размера 8 × 2



In [ ]:
a = np.linspace(30, 60, 16)
a

array([30., 32., 34., 36., 38., 40., 42., 44., 46., 48., 50., 52., 54.,
       56., 58., 60.])

Перевод в двумерный numpy-массив (вектор столбец)

In [ ]:
a.reshape(16, 1)

array([[30.],
       [32.],
       [34.],
       [36.],
       [38.],
       [40.],
       [42.],
       [44.],
       [46.],
       [48.],
       [50.],
       [52.],
       [54.],
       [56.],
       [58.],
       [60.]])

Перевод в двумерный numpy-массив (вектор-столбец) с автоматическим вычислением строк:

In [ ]:
a.reshape(-1, 1) #или np.array([a]).T

array([[30.],
       [32.],
       [34.],
       [36.],
       [38.],
       [40.],
       [42.],
       [44.],
       [46.],
       [48.],
       [50.],
       [52.],
       [54.],
       [56.],
       [58.],
       [60.]])

Используя добавление новой оси:

In [ ]:
a[np.newaxis, :].T

array([[30.],
       [32.],
       [34.],
       [36.],
       [38.],
       [40.],
       [42.],
       [44.],
       [46.],
       [48.],
       [50.],
       [52.],
       [54.],
       [56.],
       [58.],
       [60.]])

Перевод в двумерный numpy-массив размера 8 × 2:

In [ ]:
a.reshape(8, 2)

array([[30., 32.],
       [34., 36.],
       [38., 40.],
       [42., 44.],
       [46., 48.],
       [50., 52.],
       [54., 56.],
       [58., 60.]])

In [ ]:
a.reshape(8, -1)

array([[30., 32.],
       [34., 36.],
       [38., 40.],
       [42., 44.],
       [46., 48.],
       [50., 52.],
       [54., 56.],
       [58., 60.]])

### **3. Создайте два одномерных numpy-массива длинной 10. Один состоящий из единиц, другой из целых чисел, начиная с пяти. Выполните следующие задания по их конкатенации:**


*   Соедините их в один одномерный numpy-массив (несколькими способами), добавив к первому значения второго.
*   Соедините их в двумерный numpy-массив (несколькими способами) так, чтобы исходные массивы были его строками.
*   Соедините их в двумерный numpy-массив (несколькими способами) так, чтобы исходные массивы были его столбцами.







Создаём указанныее массивы:

In [ ]:
a1 = np.ones(10, dtype = 'int')
a1

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [ ]:
a2 = np.arange(5, 15)
a2

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

Присоединим к первому второй разными способами

In [ ]:
np.append(a1, a2)


array([ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  5,  6,  7,  8,  9, 10, 11,
       12, 13, 14])

In [ ]:
np.hstack((a1, a2))

array([ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  5,  6,  7,  8,  9, 10, 11,
       12, 13, 14])

Соединяем их в двумерный numpy-массив так, чтобы исходные массивы были строками:

In [ ]:
np.vstack((a1, a2))

array([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
       [ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14]])

In [ ]:
np.stack((a1, a2), axis = 0)

array([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
       [ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14]])

In [ ]:
np.row_stack((a1, a2))

/tmp/ipython-input-4007705356.py:1: DeprecationWarning: `row_stack` alias is deprecated. Use `np.vstack` directly.
  np.row_stack((a1, a2))


array([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
       [ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14]])

Соединяем в двумерный numpy-массив несколькими способами так, чтобы исходные массивы были его столбцами:

In [ ]:
np.hstack((a1.reshape(-1, 1), a2.reshape(-1, 1)))

array([[ 1,  5],
       [ 1,  6],
       [ 1,  7],
       [ 1,  8],
       [ 1,  9],
       [ 1, 10],
       [ 1, 11],
       [ 1, 12],
       [ 1, 13],
       [ 1, 14]])

In [ ]:
np.concatenate((a1.reshape(-1, 1), a2.reshape(-1, 1)), axis = 1)

array([[ 1,  5],
       [ 1,  6],
       [ 1,  7],
       [ 1,  8],
       [ 1,  9],
       [ 1, 10],
       [ 1, 11],
       [ 1, 12],
       [ 1, 13],
       [ 1, 14]])

In [ ]:
np.stack((a1, a2), axis = 1)

array([[ 1,  5],
       [ 1,  6],
       [ 1,  7],
       [ 1,  8],
       [ 1,  9],
       [ 1, 10],
       [ 1, 11],
       [ 1, 12],
       [ 1, 13],
       [ 1, 14]])

In [ ]:
np.column_stack((a1, a2))

array([[ 1,  5],
       [ 1,  6],
       [ 1,  7],
       [ 1,  8],
       [ 1,  9],
       [ 1, 10],
       [ 1, 11],
       [ 1, 12],
       [ 1, 13],
       [ 1, 14]])

### **4. Сгенерируйте случайный целочисленный двумерный NumPy-массив размера 7 × 3, значения которого берутся их диапазона от 0 до 30. Выполните следующрие задания:**


*   Вывелите его первые 3 строки, выведите его второй столбец в виде одномерного numpy-массива

*   Найдите для него средние значения в столбцах, сумму элементов в строках

*   Транспонируйте и выведите его размеры

*   Растяните его в одномиерный numpy-массив



Генерация массива:

In [ ]:
np.random.seed(0)
A = np.random.randint(0, 30, (7, 3))
A

array([[12, 15, 21],
       [ 0,  3, 27],
       [ 3,  7,  9],
       [19, 21, 18],
       [ 4, 23,  6],
       [24, 24, 12],
       [26,  1,  6]])

Первые 3 строки массива:

In [ ]:
A[:3]

array([[12, 15, 21],
       [ 0,  3, 27],
       [ 3,  7,  9]])

Второй столбец:

In [ ]:
A[:, 1]

array([15,  3,  7, 21, 23, 24,  1])

Среднее значения в столбцах и сумма в строках:

In [ ]:
A.mean(axis = 0)

array([12.57142857, 13.42857143, 14.14285714])

In [ ]:
A.sum(axis = 1)

array([48, 30, 19, 58, 33, 60, 33])

Транспонирование:

In [ ]:
A.T

array([[12,  0,  3, 19,  4, 24, 26],
       [15,  3,  7, 21, 23, 24,  1],
       [21, 27,  9, 18,  6, 12,  6]])

In [ ]:
A.T.shape

(3, 7)

Растяните в одномерный numpy-массив:

In [ ]:
A.ravel()

array([12, 15, 21,  0,  3, 27,  3,  7,  9, 19, 21, 18,  4, 23,  6, 24, 24,
       12, 26,  1,  6])

In [ ]:
A.flatten()

array([12, 15, 21,  0,  3, 27,  3,  7,  9, 19, 21, 18,  4, 23,  6, 24, 24,
       12, 26,  1,  6])

### **5. Даны две матрицы (два двумерных NumPy массива) А и В. Решите матричное уравнение AX = B.**

In [ ]:
A= np.array([[1, 2, 3], [4, 5, 6], [8, 9, 11]])

# В исходном тексте лабораторной была несколько другая матрица, а именно:
# array([[ 1,  2,  3],
#        [ 4,  5,  6],
#        [ 7,  8, 9]])
# Но из-за того, что матрица получается вырожденная, функция выдает ошибку.
# Для решения данной проблемы, было принято решение несколько изменить матрицу,
# для получения и рабочей программы и
# "хорошего" ответа.

A

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 8,  9, 11]])

In [ ]:
B = np.ones(A.shape, dtype = 'int')
B

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]])

In [ ]:
np.linalg.solve(A,B)

array([[-1., -1., -1.],
       [ 1.,  1.,  1.],
       [-0., -0., -0.]])

## **Библиотека pandas**

### **6. Скачайте с kaggle.com [датасет о пингвинах](https://www.kaggle.com/datasets/satyajeetrai/palmer-penguins-dataset-for-eda). Считайте его корректно в  pandas DataFrame и выведите первые 5 строк. Создайте копию датафрейма и далее работайте с копией. Выведите случайные 5 строк, последние З строки.**

Считываем данные и выводим первые 5 строк:

In [ ]:
data = pd.read_csv('/kaggle/input/palmer-penguins-dataset-for-eda/penguins.csv')
data.head()

id species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0   0  Adelie  Torgersen            39.1           18.7              181.0   
1   1  Adelie  Torgersen            39.5           17.4              186.0   
2   2  Adelie  Torgersen            40.3           18.0              195.0   
3   3  Adelie  Torgersen             NaN            NaN                NaN   
4   4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  year  
0       3750.0    male  2007  
1       3800.0  female  2007  
2       3250.0  female  2007  
3          NaN     NaN  2007  
4       3450.0  female  2007

Данные считаны не совсем корректно, т.к. первый столбец представляет собой индексы строк, а датафрейм тоже сам индексирует строки. Считаем данные так, чтобы первая колонка и была колонкой с индексами:

In [ ]:
data = pd.read_csv('/kaggle/input/palmer-penguins-dataset-for-eda/penguins.csv', index_col = 0)
data.head(3)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                        
0   Adelie  Torgersen            39.1           18.7              181.0   
1   Adelie  Torgersen            39.5           17.4              186.0   
2   Adelie  Torgersen            40.3           18.0              195.0   

    body_mass_g     sex  year  
id                             
0        3750.0    male  2007  
1        3800.0  female  2007  
2        3250.0  female  2007

Создадим копию датафрейма и выведим случайные 5 строк.

In [ ]:
df = data.copy()
df.sample(5)

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                      
186  Gentoo  Biscoe            49.1           14.8              220.0   
255  Gentoo  Biscoe            49.1           15.0              228.0   
45   Adelie   Dream            39.6           18.8              190.0   
26   Adelie  Biscoe            40.6           18.6              183.0   
216  Gentoo  Biscoe            45.8           14.2              219.0   

     body_mass_g     sex  year  
id                              
186       5150.0  female  2008  
255       5500.0    male  2009  
45        4600.0    male  2007  
26        3550.0    male  2007  
216       4700.0  female  2008

Выведем последние три строки:

In [ ]:
df.tail(3)

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                        
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
id                              
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009

### **7. Выведите столбец  body_mass_g в виде серии и в виде дата фрейма, выведите его значения в виде  numpy array. Выведите столбцы bill_length_mm и bill_depth_mm в виде отдельного датафреЙма, переведите данные из него в двумерныЙ  numpy array. Выведите вторую строку всего датафрейма в виде серии и в виде датафрейма. Выведите пятую, седьмую и предпоследнюю строки. Выведите строки с 10 по 15.**

Столбец body_mass_g в виде серии:

In [ ]:
df['body_mass_g']

id
0      3750.0
1      3800.0
2      3250.0
3         NaN
4      3450.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 344, dtype: float64

In [ ]:
type(df['body_mass_g'])

pandas.core.series.Series

Столбец body_mass_g в виде датафрейма:

In [ ]:
df[['body_mass_g']]

body_mass_g
id              
0         3750.0
1         3800.0
2         3250.0
3            NaN
4         3450.0
..           ...
339       4000.0
340       3400.0
341       3775.0
342       4100.0
343       3775.0

[344 rows x 1 columns]

In [ ]:
type(df[['body_mass_g']])

pandas.core.frame.DataFrame

Теперь столбц body_mass_g представим в numpy array и выведем несксхлько его первых значений:

In [ ]:
df['body_mass_g'].values[:7]

array([3750., 3800., 3250.,   nan, 3450., 3650., 3625.])

Выведем столбцы bill_length_mm и bill_depth_mm в виде отдельного датафрейма:

In [ ]:
df[['bill_length_mm', 'bill_depth_mm']]

bill_length_mm  bill_depth_mm
id                                
0              39.1           18.7
1              39.5           17.4
2              40.3           18.0
3               NaN            NaN
4              36.7           19.3
..              ...            ...
339            55.8           19.8
340            43.5           18.1
341            49.6           18.2
342            50.8           19.0
343            50.2           18.7

[344 rows x 2 columns]

Переведём его в numpy array и выведем 5 первых значений:

In [ ]:
df[['bill_length_mm', 'bill_depth_mm']].values
df[['bill_length_mm', 'bill_depth_mm']].values[:5]

array([[39.1, 18.7],
       [39.5, 17.4],
       [40.3, 18. ],
       [ nan,  nan],
       [36.7, 19.3]])

Выведем вторую строку всего датафрейма. В виде серии:

In [ ]:
df.iloc[1]

species                 Adelie
island               Torgersen
bill_length_mm            39.5
bill_depth_mm             17.4
flipper_length_mm        186.0
body_mass_g             3800.0
sex                     female
year                      2007
Name: 1, dtype: object

В виде датафрейма:

In [ ]:
df.iloc[[1]]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                        
1   Adelie  Torgersen            39.5           17.4              186.0   

    body_mass_g     sex  year  
id                             
1        3800.0  female  2007

Выведем пятую, седьмую и предпоследнюю строки данных:

In [ ]:
df.iloc[[4, 6, -2]]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                            
4       Adelie  Torgersen            36.7           19.3              193.0   
6       Adelie  Torgersen            38.9           17.8              181.0   
342  Chinstrap      Dream            50.8           19.0              210.0   

     body_mass_g     sex  year  
id                              
4         3450.0  female  2007  
6         3625.0  female  2007  
342       4100.0    male  2009

Выведим строки с 10 по 15:

In [ ]:
df.iloc[9:16]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                        
9   Adelie  Torgersen            42.0           20.2              190.0   
10  Adelie  Torgersen            37.8           17.1              186.0   
11  Adelie  Torgersen            37.8           17.3              180.0   
12  Adelie  Torgersen            41.1           17.6              182.0   
13  Adelie  Torgersen            38.6           21.2              191.0   
14  Adelie  Torgersen            34.6           21.1              198.0   
15  Adelie  Torgersen            36.6           17.8              185.0   

    body_mass_g     sex  year  
id                             
9        4250.0     NaN  2007  
10       3300.0     NaN  2007  
11       3700.0     NaN  2007  
12       3200.0  female  2007  
13       3800.0    male  2007  
14       4400.0    male  2007  
15       3700.0  female  2007

### **8. Выведите информацию о данных и проанализируйте её. Выведите размер датафрейма, а также отдельно количество строк и количество столбцов. Выведите названия столбцов и индексы строк в виде pandas Index и numpy array. Выведите тип данных каждого столбца. Определите количество пропусков в каждом столбце.**

Информация о данных:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 24.2+ KB


По ней видно, что в датафрейме 344 строки,  роиндексированных от 0 до 343. Можно прочитать названия колонок и тип данных в каждом столбце. Здесь 4 столбца имеют тип float64, 1 столбец типа int64 и З столбца типа object. Пропущенные значения имеются в столбцах
bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g и sex.

Размер таблицы данных (датафрейма)

In [ ]:
df.shape

(344, 8)

Количество строк:

In [ ]:
df.shape[0]

344

Количество столбцов:

In [ ]:
df.shape[1]

8

Названия столбцов в виде pandas Index:

In [ ]:
df.columns

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [ ]:
type(df.columns)

pandas.core.indexes.base.Index

Названия стобцов в виде numpy array:

In [ ]:
df.columns.values

array(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'], dtype=object)

Индексы строк с типом данных pandas Index:

In [ ]:
df.index

Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
       ...
       334, 335, 336, 337, 338, 339, 340, 341, 342, 343],
      dtype='int64', name='id', length=344)

In [ ]:
type(df.index)

pandas.core.indexes.base.Index

Индексы строк в виде numpy array:

In [ ]:
df.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

In [ ]:
df.index.to_numpy()

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Тип данных каждого столбца:

In [ ]:
df.dtypes

species               object
island                object
bill_length_mm       float64
bill_depth_mm        float64
flipper_length_mm    float64
body_mass_g          float64
sex                   object
year                   int64
dtype: object

Количество пропущенных значений в каждом столбце:

In [ ]:
df.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

### **9. Определите какие виды пингвинов имеются в датасете (столбец species) и сколько пингвинов каждого вида. Определите количество уникальных значений в каждом столбце. Выведите названия числовых и не числовых столбцов. Выедите статистики для числовых и не числовых столбцов.**

Определить какие элементы содержит серия и сколько элементов каждого вида в ней удобнее всего, используя метод .value_counts():

In [ ]:
df['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

КОличество уникальных значений в каждом столбце:

In [ ]:
df.nunique()

species                3
island                 3
bill_length_mm       164
bill_depth_mm         80
flipper_length_mm     55
body_mass_g           94
sex                    2
year                   3
dtype: int64

Название числовых столбцов:

In [ ]:
df.select_dtypes(include = 'number').columns

Index(['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g',
       'year'],
      dtype='object')

Названия не чиссловых столбцов:

In [ ]:
df.select_dtypes(exclude = 'number').columns

Index(['species', 'island', 'sex'], dtype='object')

Статистика по числовым столбцам:

In [ ]:
df.describe(include = 'number')

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [ ]:
df.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

Статистика по остальным столбцам:

In [ ]:
df.describe(include = 'object')

species  island   sex
count      344     344   333
unique       3       3     2
top     Adelie  Biscoe  male
freq       152     168   168

### **10. Выведите строки с пропущенными значениями. Удалите строки, в которых больше одного пропущенного значения. Замените все отсутствующие значения в столбце sex самым распространенным значением. Убедитесь, что пропущенных значений в датасете больше нет.**

Строки с пропущенным значением:

In [ ]:
df[df.isna().any(axis = 1)]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                         
3    Adelie  Torgersen             NaN            NaN                NaN   
8    Adelie  Torgersen            34.1           18.1              193.0   
9    Adelie  Torgersen            42.0           20.2              190.0   
10   Adelie  Torgersen            37.8           17.1              186.0   
11   Adelie  Torgersen            37.8           17.3              180.0   
47   Adelie      Dream            37.5           18.9              179.0   
178  Gentoo     Biscoe            44.5           14.3              216.0   
218  Gentoo     Biscoe            46.2           14.4              214.0   
256  Gentoo     Biscoe            47.3           13.8              216.0   
268  Gentoo     Biscoe            44.5           15.7              217.0   
271  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  year  
id                           
3            NaN  NaN  2007  
8         3475.0  NaN  2007  
9         4250.0  NaN  2007  
10        3300.0  NaN  2007  
11        3700.0  NaN  2007  
47        2975.0  NaN  2007  
178       4100.0  NaN  2007  
218       4650.0  NaN  2008  
256       4725.0  NaN  2009  
268       4875.0  NaN  2009  
271          NaN  NaN  2009

Строки, в которых больше одного отсутствующего значения:

In [ ]:
df[df.isnull().sum(axis = 1) >= 2]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
id                                                                         
3    Adelie  Torgersen             NaN            NaN                NaN   
271  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  year  
id                           
3            NaN  NaN  2007  
271          NaN  NaN  2009

Таких строк оказалось две.

Удалим строки, используя метод .dropna(), указав параметр thresh (количество non-NA
в строке) равным 7, так как всего 8 столбцов, тогда без пропусков в строке должно быть 7
значений. Также установим параметр чтобы сделать переиндексацию всех данных снова, т.к. были удалены строки:

In [ ]:
df = df.dropna(thresh=7, ignore_index=True)

Убедимся, что строк стало на 2 меньше:

In [ ]:
df.shape[0]

342

In [ ]:
df['sex'].value_counts()

sex
male      168
female    165
Name: count, dtype: int64

In [ ]:
df['sex'].value_counts(dropna=False)

sex
male      168
female    165
NaN         9
Name: count, dtype: int64

Выведем самое распространённое значение в столбце "sex":

In [ ]:
top_sex = df['sex'].value_counts().index[0]
top_sex

'male'

Заменим этим значением все отсутсвующие значения в столбце sex:

In [ ]:
df['sex'] = df['sex'].fillna(top_sex)
df['sex'].value_counts()

sex
male      177
female    165
Name: count, dtype: int64

Убедимся, что пропусков во всём датасете больше нет:

In [ ]:
df.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

### **11. Удалите столбец species. Добавьте столбец bill, как произведение значений столбцов bill_length_mm и bill_depth_mm.**

In [ ]:
df.head(3)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   

   body_mass_g     sex  year  
0       3750.0    male  2007  
1       3800.0  female  2007  
2       3250.0  female  2007

Удалим столбец, используя метод .drop():

In [ ]:
df = df.drop('species', axis = 1)
df.head(3)

island  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
0  Torgersen            39.1           18.7              181.0       3750.0   
1  Torgersen            39.5           17.4              186.0       3800.0   
2  Torgersen            40.3           18.0              195.0       3250.0   

      sex  year  
0    male  2007  
1  female  2007  
2  female  2007

Добавляем новый столбец:

In [ ]:
df['bill'] = df['bill_length_mm']*df['bill_depth_mm']
df.head(3)

island  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
0  Torgersen            39.1           18.7              181.0       3750.0   
1  Torgersen            39.5           17.4              186.0       3800.0   
2  Torgersen            40.3           18.0              195.0       3250.0   

      sex  year    bill  
0    male  2007  731.17  
1  female  2007  687.30  
2  female  2007  725.40

### **12. Переименуйте названия колонок bill_length_mm, bill_depth_mm, flipper_length_mm на bill_length, bill_depth, flipper_length. Сохраните получившийся набор данных в формат xlsx.**

Переименовываем колонки в соответсвии со словарём:

In [ ]:
df = df.rename(columns={"bill_length_mm" : "bill_length", "bill_depth_mm" : "bill_depth", "flipper_length_mm" : "flipper_length"})
df.head(3)

island  bill_length  bill_depth  flipper_length  body_mass_g     sex  \
0  Torgersen         39.1        18.7           181.0       3750.0    male   
1  Torgersen         39.5        17.4           186.0       3800.0  female   
2  Torgersen         40.3        18.0           195.0       3250.0  female   

   year    bill  
0  2007  731.17  
1  2007  687.30  
2  2007  725.40

Сохраним даннные в формате xlsx:

In [ ]:
df.to_excel('new_penguins.xlsx')

### **13. Считайте корректно сохранённый в предыдущем задании файл в новый датафрейм. Определите количество пингвинов мужского пола, имеющих массу тела более 4500 и встречающихся на острове Biscoe.**

In [ ]:
pd.read_excel('new_penguins.xlsx').head(3)

Unnamed: 0     island  bill_length  bill_depth  flipper_length  \
0           0  Torgersen         39.1        18.7             181   
1           1  Torgersen         39.5        17.4             186   
2           2  Torgersen         40.3        18.0             195   

   body_mass_g     sex  year    bill  
0         3750    male  2007  731.17  
1         3800  female  2007  687.30  
2         3250  female  2007  725.40

Считаем корректно:

In [ ]:
new_df = pd.read_excel('new_penguins.xlsx', index_col=0)
new_df.head(3)

island  bill_length  bill_depth  flipper_length  body_mass_g     sex  \
0  Torgersen         39.1        18.7             181         3750    male   
1  Torgersen         39.5        17.4             186         3800  female   
2  Torgersen         40.3        18.0             195         3250  female   

   year    bill  
0  2007  731.17  
1  2007  687.30  
2  2007  725.40

Выведем строки с пингвинами мужского пола:

In [ ]:
new_df[new_df['sex'] == 'male']

island  bill_length  bill_depth  flipper_length  body_mass_g   sex  \
0    Torgersen         39.1        18.7             181         3750  male   
4    Torgersen         39.3        20.6             190         3650  male   
6    Torgersen         39.2        19.6             195         4675  male   
7    Torgersen         34.1        18.1             193         3475  male   
8    Torgersen         42.0        20.2             190         4250  male   
..         ...          ...         ...             ...          ...   ...   
332      Dream         50.2        18.8             202         3800  male   
334      Dream         51.9        19.5             206         3950  male   
337      Dream         55.8        19.8             207         4000  male   
339      Dream         49.6        18.2             193         3775  male   
340      Dream         50.8        19.0             210         4100  male   

     year     bill  
0    2007   731.17  
4    2007   809.58  
6    2007   768.32  
7    2007   617.21  
8    2007   848.40  
..    ...      ...  
332  2009   943.76  
334  2009  1012.05  
337  2009  1104.84  
339  2009   902.72  
340  2009   965.20  

[177 rows x 8 columns]

Их количество:

In [ ]:
new_df[new_df['sex'] == 'male'].shape[0]

177

Определим количество пингвинов мужского пола, имеющих массу тела более 4500 и встречаются на острове Biscoe:

In [ ]:
new_df[
    (new_df['sex'] == 'male') &
    (new_df['body_mass_g'] > 4500) &
    (new_df['island'] == 'Biscoe')
      ].shape[0]

67